languages/scala/recursive_descent/Parser_scala_recursive_descent.ml

```diff
@@ -2303,9 +2303,17 @@ and interpolatedString ~inPattern in_ =
          * that?
          *)
         | ID_DOLLAR _ ->
-            let x = ident in_ in
+            let (s, t) as x = ident in_ in
             (* ast: Ident(x) *)
-            xs += EncapsDollarIdent x
+            begin
+                (* if the identifier starts with $, it is a metavariable, not a variable
+                 * inside an interpolated string. However, if we are not in semgrep_mode,
+                 * we treat these identifiers as normal variables, so we strip the $ *)
+                match s.[0], Domain.DLS.get Flag_parsing.sgrep_mode with
+                | '$', true -> xs += EncapsExpr (Name (Id x, []))
+                | '$', false -> xs += EncapsDollarIdent (String.sub s 1 (String.length s - 1), t)
+                | _ -> xs += EncapsDollarIdent x
```

Comment:
It would probably be more efficient to check `Char.equal s.[0] '$'` and only in that case read from DLS.